pkg/kv/kvserver/replica_application_result.go

```diff
@@ -225,6 +280,160 @@ func (r *Replica) prepareLocalResult(ctx context.Context, cmd *replicatedCmd) {
+	// latch.
+	var success bool
+
+	// Go through the original proposal field by field and decide what transfers
+	// to the new proposal (and how that affects the old proposal). The overall
+	// goal is that the old proposal remains a local proposal (switching it to
+	// non-local now invites logic bugs) but not bound to the caller.
+
+	// NB: quotaAlloc is always nil here, because we already
+	// released the quota unconditionally in retrieveLocalProposalsV2.
+	// So the below is a no-op.
+	//
+	// TODO(tbg): if we shifted the release of proposal quota to *after*
+	// successful application, we could move the quota over
+	// prematurely releasing it here.
+	newQuotaAlloc := origP.quotaAlloc
+	defer func() {
+		if success {
+			origP.quotaAlloc = nil
+		}
+	}()
+
+	newCommand := kvserverpb.RaftCommand{
+		ProposerLeaseSequence:   origP.command.ProposerLeaseSequence,
+		DeprecatedProposerLease: origP.command.DeprecatedProposerLease,
+		ReplicatedEvalResult:    origP.command.ReplicatedEvalResult,
+		WriteBatch:              origP.command.WriteBatch,
+		LogicalOpLog:            origP.command.LogicalOpLog,
+		TraceData:               origP.command.TraceData,
+
+		MaxLeaseIndex:       0,   // assigned on flush
+		ClosedTimestamp:     nil, // assigned on flush
+		AdmissionPriority:   0,   // assigned on flush
+		AdmissionCreateTime: 0,   // assigned on flush
+		AdmissionOriginNode: 0,   // assigned on flush
+	}
+
+	// Now we construct the remainder of the ProposalData. First, the pieces
+	// that actively "move over", i.e. those that have to do with the latches
+	// held and the caller waiting to be signaled.
+
+	// `ec` (latches, etc) transfers to the new proposal.
+	newEC := origP.ec
+	defer func() {
+		if success {
+			origP.ec = endCmds{}
+		}
+	}()
+
+	// Ditto doneCh (signal to proposer).
+	newDoneCh := origP.doneCh
+	defer func() {
+		if success {
+			origP.doneCh = nil
+		}
+	}()
+
+	r.mu.RLock()
+	ticks := r.mu.ticks
+	r.mu.RUnlock()
+
+	// TODO(tbg): work on the lifecycle of ProposalData. This struct (and the
+	// surrounding replicatedCmd) are populated in an overly ad-hoc manner.
+	// TODO(tbg): the fields are spelled out here to make explicit what is being copied
+	// here. Add a unit test that fails on addition of a new field and points at the
+	// need to double check what the intended behavior of the new field in this method
+	// is.
+	newProposal := &ProposalData{
+		// The proposal's context and span carry over. Recall that they are *not*
+		// used for command application; `cmd.{ctx,sp}` are; and since this last
+		// span "follows from" the proposal's span, if the proposal sticks around
+		// for (some reincarnation of) the command to eventually apply, its trace
+		// will reflect the reproposal as well.
+		ctx:             origP.ctx,
+		sp:              origP.sp, // NB: special handling below
+		idKey:           raftlog.MakeCmdIDKey(),
+		proposedAtTicks: ticks,
+		createdAtTicks:  ticks,
```

Comment:
We can get rid of one of these `AtTicks` fields once this migration is fully complete, right?

Reply:
Why? The purpose of these fields is to let us determine the age of a proposal (to determine whether to repropose it in `refreshProposalsLocked`: https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvserver/replica_raft.go#L1380

Reply:
Hm, right. Shouldn't we carry over `createdAtTicks` from the original proposal in that case? Or is the idea that we got an actual error result on the previous proposal, so this is a new proposal for all intents and purposes?

Since we use this for observability, and users presumably only care about how long the original request takes, it seems to make sense to retain the original `createdAtTicks`.

Reply:
Small correction, `createdAtTicks` is used for the circuit breaker, but if we get proposals through raft and pick new lease applied indexes, I don't think that should end up circuit-breaking the replica (though it sure would be odd if it happened for 1000s of round-trips for a single user command...).
The refresh mechanism only uses `proposedAtTicks`.

> Since we use this for observability, and users presumably only care about how long the original request takes, it seems to make sense to retain the original createdAtTicks.

See above, I am leaning the other way for this PR. It can go either way, honestly, but I'd rather introduce "proposal-spanning" properties deliberately.

Where do we use this for observability? Ticks aren't a great unit of measurement for user-facing purposes anyway since it's so discrete. We could change `proposedAtTicks` into a `time.Time`, then keep it across reproposals, and then when the proposal is signaled report a metric (user-visible replication latency). For our internal latency metric, we'd want to measure each individual round-trip through the replication layer, though, which would mean adding another timestamp that _is_ reset on reproposals. I think we'd best do things like that separately though, in the context of https://github.com/cockroachdb/cockroach/issues/83262.

Reply:
I thought we used it to detect slow proposals, but you're right that it's only used for the circuit breaker. This seems fine then.